src/BlazorWebView/src/Maui/Windows/WinUIWebViewManager.cs

```diff
@@ -11,11 +16,68 @@ namespace Microsoft.AspNetCore.Components.WebView.Maui
 	/// </summary>
 	public class WinUIWebViewManager : WebView2WebViewManager
 	{
-		readonly WebView2Control _nativeWebView2;
+		private readonly WebView2Control _nativeWebView2;
+		private readonly string _hostPageRelativePath;
+		private readonly string _contentRootDir;
 
-		public WinUIWebViewManager(WebView2Control nativeWebView2, IWebView2Wrapper webview, IServiceProvider services, Dispatcher dispatcher, IFileProvider fileProvider, string hostPageRelativePath) : base(webview, services, dispatcher, fileProvider, hostPageRelativePath)
+		public WinUIWebViewManager(WebView2Control nativeWebView2, IWebView2Wrapper webview, IServiceProvider services, Dispatcher dispatcher, IFileProvider fileProvider, string hostPageRelativePath, string contentRootDir) : base(webview, services, dispatcher, fileProvider, hostPageRelativePath)
 		{
 			_nativeWebView2 = nativeWebView2;
+			_hostPageRelativePath = hostPageRelativePath;
+			_contentRootDir = contentRootDir;
+		}
+
+		protected override void HandleWebResourceRequest(ICoreWebView2WebResourceRequestedEventArgsWrapper eventArgs)
+		{
+			// Unlike server-side code, we get told exactly why the browser is making the request,
+			// so we can be smarter about fallback. We can ensure that 'fetch' requests never result
+			// in fallback, for example.
+			var allowFallbackOnHostPage =
+				eventArgs.ResourceContext == CoreWebView2WebResourceContextWrapper.Document ||
+				eventArgs.ResourceContext == CoreWebView2WebResourceContextWrapper.Other; // e.g., dev tools requesting page source
+
+			// First, call into WebViewManager to see if it has a framework file for this request. It will
+			// fall back to an IFileProvider, but on WinUI it's always a NullFileProvider, so that will never
+			// return a file.
+			if (TryGetResponseContent(eventArgs.Request.Uri, allowFallbackOnHostPage, out var statusCode, out var statusMessage, out var content, out var headers)
+				&& statusCode != 404)
+			{
+				// NOTE: This is stream copying is to work around a hanging bug in WinRT with managed streams
+				var memStream = new MemoryStream();
+				content.CopyTo(memStream);
+				var ms = new InMemoryRandomAccessStream();
+				ms.WriteAsync(memStream.GetWindowsRuntimeBuffer()).AsTask().Wait();
+
+				var headerString = GetHeaderString(headers);
+				eventArgs.SetResponse(ms, statusCode, statusMessage, headerString);
+			}
+			else
+			{
+				// Next, try to go through WinUI Storage to find a static web asset
+				var uri = new Uri(eventArgs.Request.Uri);
+				if (new Uri(AppOrigin).IsBaseOf(uri))
+				{
+					var relativePath = new Uri(AppOrigin).MakeRelativeUri(uri).ToString();
+					if (allowFallbackOnHostPage && string.IsNullOrEmpty(relativePath))
+					{
+						relativePath = _hostPageRelativePath;
+					}
+					relativePath = Path.Combine("Assets", _contentRootDir, relativePath.Replace("/", "\\"));
```

Comment:
Just a note that this maybe should be a constant for now. I know there might be plans to make the images and assets go to other folders instead of just Assets. With Android, there is no choice, but since Windows might allow it, there is always someone that just has to use another folder.

Also, would this generally be configurable for all platforms because it may be the case that someone wants to put the web assets in Assets\wwwroot, and then have other, non-web stuff in Assets\Music?

Reply:
However, this is more for a later PR. Not sure about the desired configurability.

Reply:
So ideally I would love there to be a runtime MauiAssets type that could do this for me. Was there ever any discussion on some kind of `MauiAsset.GetItem(string path)` that would work on all platforms?

Reply:
cc @redth for this too.